### ISSUE TITLE
Bilibili.tv download of subtitles not possible to error Unexpected UTF-8 BOM (decode using utf-8-sig)

### ISSUE THREAD

**Issue by quamt** [yt-dlp/yt-dlp#6664]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [ ] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

When trying to download an episode from Bilibili.tv, certain subtitles can not be downloaded due to an error message:
`Unexpected UTF-8 BOM (decode using utf-8-sig)`

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', '--sub-format', 'best', '--sub-langs', 'en*', '--write-subs', '--cookies-from-browser', 'chrome', 'https://www.bilibili.tv/en/play/2071979/12664419?bstar_from=bstar-web.pgc-video-detail.episode.all']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.03.04 [392389b7d] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.22621-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg N-110112-gd6e30291db-20230328 (setts), ffprobe N-110112-gd6e30291db-20230328
[debug] Optional libraries: Cryptodome-3.17, brotli-1.0.9, certifi-2022.12.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-10.4
[Cookies] Extracting cookies from chrome
[debug] Extracting cookies from: "C:\Users\RED\AppData\Local\Google\Chrome\User Data\Default\Network\Cookies"
[debug] Found local state file at "C:\Users\RED\AppData\Local\Google\Chrome\User Data\Local State"
[Cookies] Extracted 1595 cookies from chrome
[debug] cookie version breakdown: {'v10': 1595, 'other': 0, 'unencrypted': 0}
[debug] Proxy map: {}
[debug] Loaded 1786 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Available version: stable@2023.03.04, Current version: stable@2023.03.04
Current Build Hash: 5590c57bd0433ed239a2deaaf92e2ad6f37fe50f53664c821575cafe106a9421
yt-dlp is up to date (stable@2023.03.04)
[BiliIntl] Extracting URL: https://www.bilibili.tv/en/play/2071979/12664419?bstar_from=bstar-web.pgc-video-detail.episode.all
[BiliIntl] 12664419: Downloading JSON metadata
[BiliIntl] 12664419: Downloading webpage
[BiliIntl] 12664419: Downloading JSON metadata
[BiliIntl] 12664419: Downloading video formats
[BiliIntl] 12664419: Downloading subtitles list
[BiliIntl] 12664419: Downloading subtitles for English
WARNING: [BiliIntl] 12664419: Failed to parse JSON: Unexpected UTF-8 BOM (decode using utf-8-sig): line 1 column 1 (char 0)
[BiliIntl] 12664419: Downloading subtitles for ภาษาไทย
WARNING: [BiliIntl] 12664419: Failed to parse JSON: Unexpected UTF-8 BOM (decode using utf-8-sig): line 1 column 1 (char 0)
[BiliIntl] 12664419: Downloading subtitles for Tiếng Việt
[BiliIntl] 12664419: Downloading subtitles for Bahasa Indonesia
WARNING: [BiliIntl] 12664419: Failed to parse JSON: Unexpected UTF-8 BOM (decode using utf-8-sig): line 1 column 1 (char 0)
[BiliIntl] 12664419: Downloading subtitles for Bahasa Melayu
[BiliIntl] 12664419: Downloading subtitles for 中文（简体）
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 12664419: Downloading 1 format(s): 10+2
[info] There's no subtitles for the requested languages
[debug] Invoking http downloader on "https://upos-sz-mirrorgcsbstar1.bilivideo.com/iupxcodeboss/g3/pb/n230217er1wb5k3yoxfn7b2bw53npbg3-1-281210110000.m4s?e=ig8euxZM2rNcNbdlhoNvNC8BqJIzNbfqXBvEqxTEto8BTrNvN0GvT90W5JZMkX_YN0MvXg8gNEV4NC8xNEV4N03eN0B5tZlqNxTEto8BTrNvNeZVuJ10Kj_g2UB02J0mN0B5tZlqNCNEto8BTrNvNC7MTX502C8f2jmMQJ6mqF2fka1mqx6gqj0eN0B599M=&uipk=5&nbs=1&deadline=1680029849&gen=playurlv2&os=gcsbstar1bv&oi=1951827627&trid=1445ee3b922e49828ef9665fe4ea8682i&mid=1382490330&platform=pc&upsig=274d7aa2ae34480c802dbaac21b56b00&uparams=e,uipk,nbs,deadline,gen,os,oi,trid,mid,platform&URLPrefix=aHR0cHM6Ly91cG9zLXN6LW1pcnJvcmdjc2JzdGFyMS5iaWxpdmlkZW8uY29tL2l1cHhjb2RlYm9zcy9nMy9wYi9uMjMwMjE3ZXIxd2I1azN5b3hmbjdiMmJ3NTNucGJnMy0xLTI4MTIxMDExMDAwMC5tNHM=&Expires=1680029849&KeyName=bilibili&Signature=9CdmFK6AGZUp1zTWV0S_MlSLa1k=&bvc=vod&nettype=0&orderid=0,2&logo=00000000"
[debug] File locking is not supported. Proceeding without locking
[download] Destination: E11 - Dariel Confronts [12664419].f10.mp4
[download] 100% of  831.90MiB in 00:00:15 at 52.72MiB/s
[debug] Invoking http downloader on "https://upos-sz-mirrorgcsbstar1.bilivideo.com/iupxcodeboss/g3/pb/n230217er1wb5k3yoxfn7b2bw53npbg3-1-2d1301000023.m4s?e=ig8euxZM2rNcNbdlhoNvNC8BqJIzNbfqXBvEqxTEto8BTrNvN0GvT90W5JZMkX_YN0MvXg8gNEV4NC8xNEV4N03eN0B5tZlqNxTEto8BTrNvNeZVuJ10Kj_g2UB02J0mN0B5tZlqNCNEto8BTrNvNC7MTX502C8f2jmMQJ6mqF2fka1mqx6gqj0eN0B599M=&uipk=5&nbs=1&deadline=1680029849&gen=playurlv2&os=gcsbstar1bv&oi=1951827627&trid=1445ee3b922e49828ef9665fe4ea8682i&mid=1382490330&platform=pc&upsig=827f32610eb9c6c07a0acfb783c245aa&uparams=e,uipk,nbs,deadline,gen,os,oi,trid,mid,platform&URLPrefix=aHR0cHM6Ly91cG9zLXN6LW1pcnJvcmdjc2JzdGFyMS5iaWxpdmlkZW8uY29tL2l1cHhjb2RlYm9zcy9nMy9wYi9uMjMwMjE3ZXIxd2I1azN5b3hmbjdiMmJ3NTNucGJnMy0xLTJkMTMwMTAwMDAyMy5tNHM=&Expires=1680029849&KeyName=bilibili&Signature=h5SfBqcjUkzEfBDWuWOh1d8lOow=&bvc=vod&nettype=0&orderid=0,2&logo=00000000"
[download] Destination: E11 - Dariel Confronts [12664419].f2.mp4
[download] 100% of   32.47MiB in 00:00:00 at 53.25MiB/s
[Merger] Merging formats into "E11 - Dariel Confronts [12664419].mp4"
[debug] ffmpeg command line: ffmpeg -y -loglevel "repeat+info" -i "file:E11 - Dariel Confronts [12664419].f10.mp4" -i "file:E11 - Dariel Confronts [12664419].f2.mp4" -c copy -map "0:v:0" -map "1:a:0" -movflags "+faststart" "file:E11 - Dariel Confronts [12664419].temp.mp4"
Deleting original file E11 - Dariel Confronts [12664419].f2.mp4 (pass -k to keep)
Deleting original file E11 - Dariel Confronts [12664419].f10.mp4 (pass -k to keep)
```

**Comment by rapdodge** [comment]
Is there any chances to download directly the subtitle?

Lately, bstation are having .ass subtitle natively

```https://api.bilibili.tv/intl/gateway/web/v2/subtitle?s_locale=id_ID&platform=web&episode_id=12662281```

![image](https://user-images.githubusercontent.com/33513626/229323097-2671d1d7-21e6-45e0-9a33-b1ca0c9ef931.png)

I believe yt-dlp are using the .json file
![image](https://user-images.githubusercontent.com/33513626/229323114-ae1e03d9-71eb-40d3-83ca-669651953dd5.png)

```
yt-dlp 'https://www.bilibili.tv/en/play/2072182/12662471' --list-subs -v
[debug] Command-line config: ['https://www.bilibili.tv/en/play/2072182/12662471', '--list-subs', '-v']
[debug] Portable config "/usr/local/bin/yt-dlp.conf": ['-f', 'bestvideo[height<=?720][fps<=?30][ext=webm]+bestaudio/bestvideo[height<=?720][fps<=?30][ext=mp4]+bestaudio', '--sub-langs', 'id,en', '--embed-subs', '--compat-options', 'no-live-chat', '--http-chunk-size', '10M', '-o', '%(playlist_index)s. %(title)s.%(ext)s', '-4']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.03.04 [392389b7d] (zip)
[debug] Compatibility options: no-live-chat
[debug] Python 3.9.2 (CPython x86_64 64bit) - Linux-6.1.15-1-pve-x86_64-with-glibc2.31 (OpenSSL 1.1.1n  15 Mar 2022, glibc 2.31)
[debug] exe versions: ffmpeg N-110091-g261fb55e39-20230326 (setts), ffprobe N-110091-g261fb55e39-20230326
[debug] Optional libraries: certifi-2020.06.20, no_Cryptodome-None, sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Loaded 1786 extractors
[BiliIntl] Extracting URL: https://www.bilibili.tv/en/play/2072182/12662471
[BiliIntl] 12662471: Downloading JSON metadata
[BiliIntl] 12662471: Downloading webpage
[BiliIntl] 12662471: Downloading JSON metadata
[BiliIntl] 12662471: Downloading video formats
[BiliIntl] 12662471: Downloading subtitles list
[BiliIntl] 12662471: Downloading subtitles for English
WARNING: [BiliIntl] 12662471: Failed to parse JSON: Unexpected UTF-8 BOM (decode using utf-8-sig): line 1 column 1 (char 0)
[BiliIntl] 12662471: Downloading subtitles for ภาษาไทย
WARNING: [BiliIntl] 12662471: Failed to parse JSON: Unexpected UTF-8 BOM (decode using utf-8-sig): line 1 column 1 (char 0)
[BiliIntl] 12662471: Downloading subtitles for Tiếng Việt
WARNING: [BiliIntl] 12662471: Failed to parse JSON: Unexpected UTF-8 BOM (decode using utf-8-sig): line 1 column 1 (char 0)
[BiliIntl] 12662471: Downloading subtitles for Bahasa Indonesia
WARNING: [BiliIntl] 12662471: Failed to parse JSON: Unexpected UTF-8 BOM (decode using utf-8-sig): line 1 column 1 (char 0)
[BiliIntl] 12662471: Downloading subtitles for Bahasa Melayu
WARNING: [BiliIntl] 12662471: Failed to parse JSON: Unexpected UTF-8 BOM (decode using utf-8-sig): line 1 column 1 (char 0)
[BiliIntl] 12662471: Downloading subtitles for 中文（简体）
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[info] Available subtitles for 12662471:
Language Formats
zh-Hans  srt
```

**Comment by dirkf** [comment]
In the original log there are two problems. The subtitles can be extracted without OP's login by skipping the video extraction in the debugger.

1. The subtitle text (actually only some of them), perhaps written in Windows Notepad, has to be decoded as `utf-8-sig`, per the `WARNING`.

2. The subtitles are in ASS (Unicode SSA) format while the extractor incorrectly expects JSON from a URL like https://s.bstarstatic.com/ogv/subtitle/941f876a21a90ffbbb7a75bfb68b6522.ass?auth_key=1680610089-0-0-c3190dd9ed89e7ecde77f7984520f34e.

```diff
--- /tmp/meld-tmpzi1xsj9m
+++ /home/df/Documents/src/yt-dlp/yt_dlp/extractor/bilibili.py
@@ -12,6 +12,7 @@
     GeoRestrictedError,
     InAdvancePagedList,
     OnDemandPagedList,
+    determine_ext,
     filter_dict,
     float_or_none,
     format_field,
@@ -835,14 +836,26 @@
             sub_url = sub.get('url')
             if not sub_url:
                 continue
-            sub_data = self._download_json(
-                sub_url, ep_id or aid, errnote='Unable to download subtitles', fatal=False,
-                note='Downloading subtitles%s' % f' for {sub["lang"]}' if sub.get('lang') else '')
+            ext = determine_ext(sub_url)
+            msg = (
+                'Unable to download subtitles',
+                'Downloading subtitles' + f' for {sub["lang"]}' if sub.get('lang') else '')
+            if ext == 'ass':
+                sub_data = self._download_webpage(
+                    sub_url, ep_id or aid, errnote=msg[0], fatal=False,
+                    encoding='utf-8-sig', note=msg[1])
+            else:
+                # maybe consider other types?
+                sub_data = self._download_json(
+                    sub_url, ep_id or aid, errnote=msg[0], fatal=False, note=msg[1])
+                if sub_data:
+                    sub_data = self.json2srt(sub_data)
+                    ext = 'srt'
             if not sub_data:
                 continue
             subtitles.setdefault(sub.get('lang_key', 'en'), []).append({
-                'ext': 'srt',
-                'data': self.json2srt(sub_data)
+                'ext': ext,
+                'data': sub_data
             })
         return subtitles
 
```
Then here's `[(k, subtitles[k][0]['ext']) for k in subtitles]`:
```py
[('en', 'ass'),
 ('th', 'ass'),
 ('vi', 'srt'),
 ('id', 'ass'),
 ('ms', 'srt'),
 ('zh-Hans', 'srt')]
```
